### DISCUSSION TITLE
Am I doing this right? Trying to edit Max device to receive CC from OpCo, then saving the edited device to retain CC set-up

### DISCUSSION THREAD

**Discussion by ll-dasein-ll** [KBLiveSolutions/open.control#20] [Troubleshooting] [answered]
Before I start, I really hope I can explain this clearly and that I'm asking in the right place, as I think this may be more of a Max for Live question than an OpenControl question.  But here goes:

I have a Max for Live device that has a single function I want to be able to toggle on and off with a CC command from OpenControl, no matter whether the device is selected or not - effectively, I want to MIDI map the button, but so that this single button control is saved with the device, not as part of the Live session.  This would mean I can drop the device into any project, new or old, and have OpenControl toggle the button on/off without MIDI mapping it every time.  (Just for info, the button I want to trigger on the device opens a floating window for the currently selected channel's mixer).

I've managed to make some progress with this - I've added a "ctlin" object to the device with the relevant MIDI channel details etc., connected it to the button inside the device, saved it, and it works in any project I add it to, but only partly.  I can't seem to get it to toggle on and off.  At the moment, the floating window opens when I click the button on my OpenControl, but pressing the button a second time doesn't close it again - it just moves the floating window to wherever the cursor is onscreen at that moment (which is the expected behaviour when the window opens, so it's as if the button is only ever sending "on" commands without sending an "off" command between "on" commands).  I've used a MIDI monitor app to check what's going on in the background and can see that OpenControl is correctly cycling between 127 and 0 values on each press, so I'm guessing the problem must be with the ctlin object and how I've added the variables to the object?

Is this something anyone here could help with?  I'm not sure if I'm going about this the right way (is ctlin the best solution?), or if it's even possible to get working the way that I'm envisaging, but the fact I've got so far with it makes me think it should work and it's just that I'm not good enough with Max :-).

I can send screenshots and more details of what I've tried so far, if needed (like adding the 0 127 variables to the ctlin object, but I think I'm probably doing this bit wrong, which is likely to be why I can't get this to work properly, haha).

Sorry if this isn't the right place to ask about this kind of thing, but thanks so much if anyone can help :-).

Cheers,
Martin

**Comment by KBLiveSolutions** [comment, marked as answer]
Hey Martin

One way to do that is to have a [sel 127] connected to the ctlin, and a [toggle] connected to the [sel].
This will make the CC toggle.
I hope that's clear.

I might add a device detection system similar to what I use for Loopers and Setlist, ie adding (DEVICE1) to the name of a device would add it to a virtual device list for which you can have a number of actions.

**Reply by ll-dasein-ll** [reply]
Hey Pierre -

Thank you for the advice, and yes, it's clear as crystal - I'll try it tonight and let you know how it goes :-)

Your device detection idea would be amazing - that would open up so many possibilities for live performances.  I've nearly got two pages fully mapped to my preferences for overall project control on the OpenControl now, but being able to use page 3 as a set of virtual hard-wired controls for my favourite devices would be a total game-changer!

**Comment by ll-dasein-ll** [comment]
I've only just had a chance to try this out this morning, but it's worked perfectly :-).  The only problem (which isn't a problem at all, really), is that the device needs to be on a MIDI track rather than on the Master, which is where I would usually place devices such as this.  But that's no big deal at all.  (EDIT: I've just discovered that the MIDI monitor on the channel hosting the device has to be set to _IN_ if you want it to be able to trigger regardless of which track is selected.

For anyone  else who may want to customise a Max for Live device to allow OpenControl to toggle a button on/off with a custom MIDI CC command, here's a screenshot of the simple addition I made to the device.  Just make sure you've set the MIDI channel on the OpenControl button to match the number you've typed into the _ctlin_ object and you should be good to go!:

![Screenshot 2022-05-13 at 09 56 53](https://user-images.githubusercontent.com/104435068/168270002-cbfecf00-7577-4aee-8834-7d5576e056f8.png)

Thanks again for your help with this Pierre - this little device's value is increasing almost by the day for me :-)
